### DISCUSSION TITLE
Slow convergence of plasticity model

### DISCUSSION THREAD

**Discussion by tairoon1** [idaholab/moose#16905] [Q&A Modules: Solid mechanics] [answered]
Hi everyone,

I have been trying to use tensor mechanics to simulate a plasticity example, but somehow experience extremely slow convergence using Newton solver. I tried to make a minimal example, can someone please take a look at my input file, [here](https://github.com/tairoon1/moosefiles)?

I have two materials, the matrix has a von Mises isotropic hardening model, the fiber is linear elastic. I want to use small strain formulation. The loading is applied uniaxially until time=0.5 and then unloading until time=1. During plastic deformation, it would sometimes take over 100 nonlinear iterations for Newton to converge and there is definitely no quadratic convergence. Is that normal behavior?

Thanks in advance.

**Comment by GiudGiud** [comment]
Hello

I had a quick look at your input. The non linear relative error convergence criterion seems really small. 
Can you please have a look at [https://mooseframework.inl.gov/modules/tensor_mechanics/Convergence.html](https://mooseframework.inl.gov/modules/tensor_mechanics/Convergence.html) to see if it needs to be adjusted.
Have not worked on plasticity for a while now, do you expect the small strain formulation to be valid for a 30% displacement? (mesh is 1 by 1 and `right_displacement_x` goes to 0.3).

I don't know how much the non linear iteration count may be reduced, it depends on your problem's non-linearity. To mitigate this cost, you could try adaptive time stepping to reduce the number of time steps, consequently reducing the number of non-linear solves. And you could add pre-conditionning to make your linear solves inverting the Jacobian faster. 

Guillaume

**Reply by tairoon1** [reply]
Hi, thanks for your response. Actually, I am using the absolute error convergence criterion to converge (10^-10) which should be fine, I believe.

The result is definitely unphysical and wrong, but this was also meant more as an academic example. The same happens when I would take only 3% strain with a yield stress of 1/10.

Well, the thing is, even if I set the fiber to the same material as the matrix, i.e. the material is homogeneous, the slow convergence still appears, so the problem should not be very nonlinear... Another thing that I noticed is, the more plastic strain there is, the more nonlinear iterations are needed for a timestep to converge. An example of the convergence behavior is below. Can you please take a more thorough look? I do not really understand what is happening.

Time Step 36, time = 0.36, dt = 0.01
 0 Nonlinear |R| = 3.286335e-02
 1 Nonlinear |R| = 9.414443e-04
 2 Nonlinear |R| = 4.019062e-04
 3 Nonlinear |R| = 2.413215e-04
 4 Nonlinear |R| = 1.647919e-04
 5 Nonlinear |R| = 1.191983e-04
 6 Nonlinear |R| = 8.981290e-05
 7 Nonlinear |R| = 6.902610e-05
 8 Nonlinear |R| = 5.399464e-05
 9 Nonlinear |R| = 4.289745e-05
10 Nonlinear |R| = 3.447887e-05
11 Nonlinear |R| = 2.795538e-05
12 Nonlinear |R| = 2.282174e-05
13 Nonlinear |R| = 1.873410e-05
14 Nonlinear |R| = 1.545260e-05
15 Nonlinear |R| = 1.279490e-05
16 Nonlinear |R| = 1.062149e-05
17 Nonlinear |R| = 8.845436e-06
18 Nonlinear |R| = 7.387003e-06
19 Nonlinear |R| = 6.184153e-06
20 Nonlinear |R| = 5.188392e-06
21 Nonlinear |R| = 4.361392e-06
22 Nonlinear |R| = 3.672593e-06
23 Nonlinear |R| = 3.097446e-06
24 Nonlinear |R| = 2.616114e-06
25 Nonlinear |R| = 2.212475e-06
26 Nonlinear |R| = 1.873369e-06
27 Nonlinear |R| = 1.588007e-06
28 Nonlinear |R| = 1.347508e-06
29 Nonlinear |R| = 1.144539e-06
30 Nonlinear |R| = 9.730259e-07
31 Nonlinear |R| = 8.279240e-07
32 Nonlinear |R| = 7.050327e-07
33 Nonlinear |R| = 6.008464e-07
34 Nonlinear |R| = 5.124338e-07
35 Nonlinear |R| = 4.373398e-07
36 Nonlinear |R| = 3.735039e-07
37 Nonlinear |R| = 3.191944e-07
38 Nonlinear |R| = 2.729543e-07
39 Nonlinear |R| = 2.335556e-07
40 Nonlinear |R| = 1.999625e-07
41 Nonlinear |R| = 1.713000e-07
42 Nonlinear |R| = 1.468283e-07
43 Nonlinear |R| = 1.259216e-07
44 Nonlinear |R| = 1.080495e-07
45 Nonlinear |R| = 9.276236e-08
46 Nonlinear |R| = 7.967877e-08
47 Nonlinear |R| = 6.847478e-08
48 Nonlinear |R| = 5.887508e-08
49 Nonlinear |R| = 5.064552e-08
50 Nonlinear |R| = 4.358684e-08
51 Nonlinear |R| = 3.752934e-08
52 Nonlinear |R| = 3.232840e-08
53 Nonlinear |R| = 2.786069e-08
54 Nonlinear |R| = 2.402099e-08
55 Nonlinear |R| = 2.071945e-08
56 Nonlinear |R| = 1.787932e-08
57 Nonlinear |R| = 1.543500e-08
58 Nonlinear |R| = 1.333037e-08
59 Nonlinear |R| = 1.151743e-08
60 Nonlinear |R| = 9.955062e-09
61 Nonlinear |R| = 8.608056e-09
62 Nonlinear |R| = 7.446233e-09
63 Nonlinear |R| = 6.443716e-09
64 Nonlinear |R| = 5.578305e-09
65 Nonlinear |R| = 4.830943e-09
66 Nonlinear |R| = 4.185269e-09
67 Nonlinear |R| = 3.627225e-09
68 Nonlinear |R| = 3.144729e-09
69 Nonlinear |R| = 2.727393e-09
70 Nonlinear |R| = 2.366277e-09
71 Nonlinear |R| = 2.053691e-09
72 Nonlinear |R| = 1.783012e-09
73 Nonlinear |R| = 1.548535e-09
74 Nonlinear |R| = 1.345344e-09
75 Nonlinear |R| = 1.169201e-09
76 Nonlinear |R| = 1.016452e-09
77 Nonlinear |R| = 8.839422e-10
78 Nonlinear |R| = 7.689509e-10
79 Nonlinear |R| = 6.691275e-10
80 Nonlinear |R| = 5.824420e-10
81 Nonlinear |R| = 5.071402e-10
82 Nonlinear |R| = 4.417056e-10
83 Nonlinear |R| = 3.848266e-10
84 Nonlinear |R| = 3.353688e-10
85 Nonlinear |R| = 2.923503e-10
86 Nonlinear |R| = 2.549210e-10
87 Nonlinear |R| = 2.223447e-10
88 Nonlinear |R| = 1.939836e-10
89 Nonlinear |R| = 1.692849e-10
90 Nonlinear |R| = 1.477694e-10
91 Nonlinear |R| = 1.290213e-10
92 Nonlinear |R| = 1.126800e-10
93 Nonlinear |R| = 9.843259e-11
 Solve Converged!

**Reply by GiudGiud** [reply]
Plastic strain is non linear so it causing an increase in non-linear iterations is not surprising. 

You seem to be obtaining a linear convergence (roughly 10x reduction every 15 iterations). I would examine the Jacobian to see if it's exact on a smaller problem. That would reduce your convergence rate from quadratic to linear.
You can use the command line options `-snes_test_jacobian -snes_test_jacobian_view` but the problem should be really small so the output is understandable.

**Reply by srinath-chakravarthy** [reply]
I am getting non-linear convergence within 2 non-linear iterations with your input file, but using the Tensor-mechanics master action instead of the full input file you have. It is recommended to use the MasterAction in most cases i think. Here is the input file 

[Mesh]
  file = mat_fiber.msh
  # use_displaced_mesh = false
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
[]

[Functions]
  [./Exx]
    type = PiecewiseLinear
    x = '0. 0.5 1 '
    y = '0 0.3 0'
  [../]
  [./right_displacement_x]
    type = ParsedFunction
    value = Exx
    vars = 'Exx'
    vals = 'Exx'
  [../]
[]

[BCs]
  [./left_x]
    type = DirichletBC
    variable = 'disp_x'
    boundary = left
    value = 0
  [../]
  [./left_y]
    type = DirichletBC
    variable = 'disp_y'
    boundary = left
    value = 0
  [../]
  [./right_x]
    type = FunctionDirichletBC
    variable = 'disp_x'
    boundary = right
    function = right_displacement_x
    preset   = false
  [../]
  [./right_y]
    type = DirichletBC
    variable = 'disp_y'
    boundary = right
    value = 0
  [../]
[]

[Modules/TensorMechanics/Master]
  [./all]
    add_variables = true
    strain = SMALL
    incremental = true
    # volumetric_locking_correction = true
    generate_output = 'stress_xx stress_yy strain_xx strain_yy vonmises_stress hydrostatic_stress'
  [../]
[]

[Materials]

  [./elasticity_tensor_matrix]
    type = ComputeIsotropicElasticityTensor
    block = 'matrix'
    youngs_modulus = 1.0
    poissons_ratio = 0.3
  [../]
  [./stress_matrix]
    type = ComputeMultipleInelasticStress
    inelastic_models = 'isoplas'
    block = 'matrix'
  [../]
  [./isoplas]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 0.18
    hardening_constant = 0.1
    block = 'matrix'
  [../]

  [./elasticity_tensor_fiber]
    type = ComputeIsotropicElasticityTensor
    block = 'fiber'
    youngs_modulus = 1.0
    poissons_ratio = 0.1
  [../]
  [./stress_fiber]
    type = ComputeFiniteStrainElasticStress
    block = 'fiber'
  [../]
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  start_time = 0.0
  dtmin = 0.0001

  end_time = 1.0
  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-20
  nl_max_its = 200
  dt = 0.01
[]


[Outputs]
  execute_on = 'INITIAL TIMESTEP_END'
  exodus = true
  print_linear_residuals = false
[]

**Reply by tairoon1** [reply]
@srinath-chakravarthy Thanks for your response. I tried your input file, but I observe the same convergence behavior as with my own file. The first 17 timesteps are fine since there is no plastic deformation yet. The plastic deformation starts at timestep 18.

**Reply by tairoon1** [reply]
@GiudGiud Thanks for the hint. I tried it with the command.
For elastic deformation, it is perfectly consistent with finite differences:
---------- Testing Jacobian -------------
||J - Jfd||_F/||J||_F = 1.6109e-16, ||J - Jfd||_F = 1.95199e-15

However, when plastic deformation starts:
---------- Testing Jacobian -------------
  ||J - Jfd||_F/||J||_F = 0.039895, ||J - Jfd||_F = 0.483421.

So it looks like the jacobian is wrong... But I am directly using the code from TensorMechanics Module. Does that mean that the implementation of MOOSE is incorrect?

**Comment by hugary1995** [comment, marked as answer]
Try this input instead:

```
[Mesh]
  file = mat_fiber.msh
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
[]

[Functions]
  [Exx]
    type = PiecewiseLinear
    x = '0. 0.5 1 '
    y = '0 0.3 0'
  []
  [right_displacement_x]
    type = ParsedFunction
    value = Exx
    vars = 'Exx'
    vals = 'Exx'
  []
[]

[BCs]
  [left_x]
    type = DirichletBC
    variable = 'disp_x'
    boundary = left
    value = 0
  []
  [left_y]
    type = DirichletBC
    variable = 'disp_y'
    boundary = left
    value = 0
  []
  [right_x]
    type = FunctionDirichletBC
    variable = 'disp_x'
    boundary = right
    function = right_displacement_x
    preset = false
  []
  [right_y]
    type = DirichletBC
    variable = 'disp_y'
    boundary = right
    value = 0
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = SMALL
    incremental = true
    # volumetric_locking_correction = true
    generate_output = 'stress_xx stress_yy strain_xx strain_yy vonmises_stress hydrostatic_stress'
  []
[]

[Materials]
  [elasticity_tensor_matrix]
    type = ComputeIsotropicElasticityTensor
    block = 'matrix'
    youngs_modulus = 1.0
    poissons_ratio = 0.3
  []
  [stress_matrix]
    type = ComputeMultipleInelasticStress
    inelastic_models = 'isoplas'
    block = 'matrix'
  []
  [isoplas]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 0.18
    hardening_constant = 0.1
    block = 'matrix'
  []

  [elasticity_tensor_fiber]
    type = ComputeIsotropicElasticityTensor
    block = 'fiber'
    youngs_modulus = 1.0
    poissons_ratio = 0.1
  []
  [stress_fiber]
    type = ComputeFiniteStrainElasticStress
    block = 'fiber'
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu superlu_dist'

  start_time = 0.0
  dtmin = 0.0001

  end_time = 1.0
  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-20
  nl_max_its = 200
  dt = 0.01
[]

[Outputs]
  execute_on = 'INITIAL TIMESTEP_END'
  exodus = true
  print_linear_residuals = false
[]
```
The finite deformation jacobian in tensor_mechanics isn't perfect (as of now, it will improve by a lot soon!). Using PJFNK for non-AD or using Newton for AD should improve the convergence.

Some other options you could play with are the `tangent_operator` option in ComputeMultiInelasticStress and `use_finite_deform_jacobian` in the action.

**Reply by srinath-chakravarthy** [reply]
Also, 
One more thing, you are applying a displacement bc and so the amount of plastic strain per increment is increasing, which is why you are seeing an increase in the no. of iterations per increment. Try to get non constant time steps and also try to see if you can apply a velocity bc. I have had a considerable amount of success by choosing an analytical vel. bc in lieu of displacment bc.

**Reply by tairoon1** [reply]
Thanks @hugary1995. So the jacobian in tensor_mechanics is currently not computed correctly. Are you currently working on a correct and consistent jacobian implementation? Can I ask, when soon is going to be? Using PJFNK seems to work around it quite well, however each linear iteration seems to be much slower than Newton.

**Reply by tairoon1** [reply]
@srinath-chakravarthy I am sorry, I do not really understand what you mean. I am working with quasistatic conditions, so there should not be any velocity.

**Reply by hugary1995** [reply]
I'm not - but that's in our radar for a while. @bwspenc can probably comment on this.

PJFNK uses our inaccurate hand-coded jacobian for preconditioning, so linear convergence might be slow, but each iteration is also supposed to be much cheaper. See [https://mooseframework.inl.gov/source/systems/NonlinearSystem.html](https://mooseframework.inl.gov/source/systems/NonlinearSystem.html)  for more details.

**Reply by tairoon1** [reply]
Okay, thank you so much for clarifying!

**Reply by hugary1995** [reply]
There is a new system (we refer to it as "lagrangian mechanics kernels/materials" at several places). As far as I know, the current state of Jacobian accuracy is:
- some non-AD simple plasticity models in the old system now have correct Jacobians;
- most non-AD creep models in the old system have correct Jacobians;
- all AD inelastic models in the old system have correct Jacobians;
- all inelastic models in the new system have correct Jacobians, though we only added a few inelastic models so far -- migration in progress.

**Reply by Ttw0626** [reply]
sorry this borthing, I wanna ask how to imporve the convergenced speed of non-AD inelastic model (Lagragian) which I did not use build-in return-mapping ? it will take too much time (around 2 h in 3D tensile coarse mesh) compared with commercial one (5-10mins). maybe my jacobian has problem, but it will get good convergency in 2D problem.

Best regards,
TAN

**Reply by GiudGiud** [reply]
Hello

You can check your Jacobian using these tools:
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

If your Jacobian has issues you ll see a higher number of nonlinear iterations. 

If it's not that, you will want to work on the preconditioning to improve performance in most cases.

Guillaume

**Reply by Ttw0626** [reply]
Thanks for reply, Guillaume.

1)
I am going to try.

2)
Actually, I wonder what's the reason why my material jacobian [modification based on ComputeLagrangianPK2Stress()] will failure in moose but is fine in abaqus umat when the problem become a little bit complicated?
I understood the difference of the material jacobian between TM new system and abaqus umat, however, the basic StressUpdate algorithm should be same I think.

3)
This is first implementation in moose about inelastic material, if I wanna insert more nonlinear on this framework in the future, the converged problem is super-important for me, could you give me some advice or recommendations about this ?

Best regards.
TAN
